modules/caddyhttp/metrics.go

```diff
@@ -0,0 +1,215 @@
+}
+
+type metricsInstrumentedHandler struct {
+	labels       prometheus.Labels
+	statusLabels prometheus.Labels
+	mh           MiddlewareHandler
+}
+
+func newMetricsInstrumentedHandler(server, handler string, mh MiddlewareHandler) *metricsInstrumentedHandler {
+	httpMetrics.init.Do(func() {
+		initHTTPMetrics()
+	})
+
+	labels := prometheus.Labels{"server": server, "handler": handler}
+	statusLabels := prometheus.Labels{"server": server, "handler": handler, "code": "", "method": ""}
+	return &metricsInstrumentedHandler{labels, statusLabels, mh}
+}
+
+func (h *metricsInstrumentedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request, next Handler) error {
+	inFlight := httpMetrics.requestInFlight.With(h.labels)
+	inFlight.Inc()
+	defer inFlight.Dec()
+
+	statusLabels := prometheus.Labels{"method": r.Method}
+	for k, v := range h.labels {
+		statusLabels[k] = v
+	}
+
+	start := time.Now()
+
+	// This is a _bit_ of a hack - it depends on the ShouldBufferFunc always
+	// being called when the headers are written.
+	// Effectively the same behaviour as promhttp.InstrumentHandlerTimeToWriteHeader.
+	writeHeaderRecorder := ShouldBufferFunc(func(status int, header http.Header) bool {
+		statusLabels["code"] = sanitizeCode(status)
+		ttfb := time.Since(start).Seconds()
+		observeWithExemplar(statusLabels, httpMetrics.responseLatency, ttfb)
+		httpMetrics.responseLatencySummary.With(statusLabels).Observe(ttfb)
+		return false
+	})
+	wrec := NewResponseRecorder(w, nil, writeHeaderRecorder)
+	err := h.mh.ServeHTTP(wrec, r, next)
+	dur := time.Since(start).Seconds()
+	httpMetrics.requestCount.With(h.labels).Inc()
+	if err != nil {
+		httpMetrics.requestErrors.With(h.labels).Inc()
+		return err
+	}
+
+	observeWithExemplar(statusLabels, httpMetrics.requestDuration, dur)
+	httpMetrics.requestDurationSummary.With(statusLabels).Observe(dur)
+	observeWithExemplar(statusLabels, httpMetrics.requestSize, float64(computeApproximateRequestSize(r)))
+	httpMetrics.responseSize.With(statusLabels).Observe(float64(wrec.Size()))
+
+	return nil
+}
+
+func observeWithExemplar(l prometheus.Labels, o *prometheus.HistogramVec, value float64) {
+	obs := o.With(l)
+	if oe, ok := obs.(prometheus.ExemplarObserver); ok {
+		oe.ObserveWithExemplar(value, l)
+		return
+	}
+	// _should_ be a noop, but here just in case...
+	obs.Observe(value)
+}
+
+func sanitizeCode(code int) string {
+	if code == 0 {
+		return "200"
+	}
+	return strconv.Itoa(code)
+
+}
+
+// taken from https://github.com/prometheus/client_golang/blob/6007b2b5cae01203111de55f753e76d8dac1f529/prometheus/promhttp/instrument_server.go#L298
+func computeApproximateRequestSize(r *http.Request) int {
+	s := 0
+	if r.URL != nil {
+		s += len(r.URL.String())
+	}
+
+	s += len(r.Method)
+	s += len(r.Proto)
+	for name, values := range r.Header {
+		s += len(name)
+		for _, value := range values {
+			s += len(value)
+		}
+	}
+	s += len(r.Host)
+
+	// N.B. r.Form and r.MultipartForm are assumed to be included in r.URL.
```

Comment:
What is `N.B.`?

Reply:
https://en.wikipedia.org/wiki/Nota_bene

Reply:
Or just remove it altogether, a comment is by definition a note...

Reply:
I'm a bit reluctant to change this, as the function was lifted verbatim from the Promtheus client (as indicated in the function comment).

The only reason I needed to lift this function was so that I could replicate [`promhttp. InstrumentHandlerRequestSize`](https://github.com/prometheus/client_golang/blob/6007b2b5cae01203111de55f753e76d8dac1f529/prometheus/promhttp/instrument_server.go#L157) for `caddyhttp.Handler`'s signature